toolchain/check/handle_modifier.cpp

```diff
@@ -0,0 +1,142 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/handle_modifier.h"
+
+#include "toolchain/check/context.h"
+#include "toolchain/lex/token_kind.h"
+
+namespace Carbon::Check {
+
+auto HandleDeclModifierKeyword(Context& context, Parse::Node parse_node)
+    -> bool {
+  context.node_stack().Push(parse_node);
+  return true;
+}
+
+auto ValidateModifiers(Context& context, DeclModifierKeywords allowed,
+                       std::function<Parse::Node()> pop_introducer)
+    -> std::pair<DeclModifierKeywords, Parse::Node> {
+  DeclModifierKeywords found;
+  Parse::Node saw_access = Parse::Node::Invalid;
+  Parse::Node saw_other = Parse::Node::Invalid;
+  llvm::SmallVector<Parse::Node> modifier_nodes;
+  // Note that since we are reading the modifier keywords off of a stack,
+  // we get them in reverse order compared to how they appeared in the source.
+  while (
+      auto modifier_node =
+          context.node_stack()
+              .PopForSoloParseNodeIf<Parse::NodeKind::DeclModifierKeyword>()) {
+    modifier_nodes.push_back(*modifier_node);
+  }
+  Parse::Node introducer_node = pop_introducer();
+  Parse::Node returned_node =
+      modifier_nodes.empty() ? introducer_node : modifier_nodes.back();
+
+  // Reverse the order of processing the modifier nodes back to source order.
+  while (!modifier_nodes.empty()) {
+    auto modifier_node = modifier_nodes.back();
+    modifier_nodes.pop_back();
+    auto modifier_token = context.parse_tree().node_token(modifier_node);
+
+    CARBON_DIAGNOSTIC(ModifierNotAllowedOnDeclaration, Error,
+                      "Modifier not allowed.");
+    CARBON_DIAGNOSTIC(ModifierDeclarationNote, Note, "On this declaration.");
+    CARBON_DIAGNOSTIC(ModifierDuplicated, Error,
+                      "Modifier repeated on the same declaration.");
+    CARBON_DIAGNOSTIC(ModifierDuplicatedPrevious, Note,
+                      "Previously appeared here.");
+    CARBON_DIAGNOSTIC(ModifierNotAllowedWith, Error,
+                      "Modifier not allowed on the same declaration.");
+    CARBON_DIAGNOSTIC(ModifierNotAllowedWithPrevious, Note,
+                      "With this modifier.");
```

Comment:
I wonder if it'd be easier to read these as something like:

`base` repeated on declaration.
`abstract` not allowed on declaration with `base`.
`private` must appear before `base`.

With the same note:
`base` previously appeared here.

Note too, I think @zygoloid's comment on #toolchain had been to avoid "the" in diagnostics, thus a slightly different phrasing I'm suggesting.